build_tools/cmake/iree_compiler_plugin.cmake

```diff
@@ -39,71 +33,54 @@ function(iree_compiler_register_plugin)
 endfunction()
 
 # iree_compiler_configure_plugins()
 # Configures all in-tree and out-of-tree plugins based on global settings.
 function(iree_compiler_configure_plugins)
-  # Process in-tree plugins.
-  foreach(_plugin_id ${IREE_COMPILER_PLUGINS})
-    set(_plugin_src_dir "${IREE_SOURCE_DIR}/compiler/plugins/${_plugin_id}")
-    iree_compiler_add_plugin("${_plugin_id}" "${_plugin_src_dir}")
+  set(IREE_COMPILER_PLUGINS_BINARY_DIR "${IREE_BINARY_DIR}/compiler/plugins")
+  file(MAKE_DIRECTORY "${IREE_COMPILER_PLUGINS_BINARY_DIR}")
+
+  # Generate an include() in compiler/plugins/CMakeLists.txt for each plugin dir.
+  set(_contents)
+  foreach(_d ${IREE_COMPILER_BUILTIN_PLUGIN_PATHS})
+    cmake_path(ABSOLUTE_PATH _d BASE_DIRECTORY "${IREE_SOURCE_DIR}" NORMALIZE)
+    set(_f "${_d}/iree_compiler_plugin_group.cmake")
+    if(NOT EXISTS "${_f}")
+      message(SEND_ERROR "Specified plugin directory ${_d} does not contain an iree_compiler_plugin_group.cmake file")
+      continue()
+    endif()
+    string(APPEND _contents "include(${_f})\n")
   endforeach()
-  unset(_plugin_id)
-  unset(_plugin_src_dir)
-
-  # Process out of tree plugins.
-  foreach(_plugin_src_dir ${IREE_COMPILER_PLUGIN_PATHS})
-    # TODO: Support some path mangling to allow overriding the plugin id
-    # if it is not literally the last path component.
-    cmake_path(ABSOLUTE_PATH _plugin_src_dir BASE_DIRECTORY "${IREE_SOURCE_DIR}" NORMALIZE)
-    cmake_path(GET _plugin_src_dir FILENAME _plugin_id)
-    iree_compiler_add_plugin("${_plugin_id}" "${_plugin_src_dir}")
+  foreach(_d ${IREE_COMPILER_PLUGIN_PATHS})
+    cmake_path(ABSOLUTE_PATH _d BASE_DIRECTORY "${IREE_SOURCE_DIR}" NORMALIZE)
+    set(_f "${_d}/iree_compiler_plugin_group.cmake")
+    if(NOT EXISTS "${_f}")
+      message(SEND_ERROR "Specified plugin directory ${_d} does not contain an iree_compiler_plugin_group.cmake file")
+      continue()
+    endif()
+    string(APPEND _contents "include(${_f})\n")
   endforeach()
-endfunction()
-
-# iree_compiler_add_plugin(src bin)
-# Adds a compiler plugin based on id and source directory.
-function(iree_compiler_add_plugin plugin_id plugin_src_dir)
-  # Add a guard so that we know if we are adding an in-tree plugin.
-  if(IREE_COMPILER_IN_ADD_PLUGIN)
-    message(FATAL_ERROR "Cannot recursively add plugins")
-  endif()
-
-  message(STATUS "Adding static compiler plugin ${plugin_id} (from ${plugin_src_dir})")
-  get_property(_existing_plugin_ids GLOBAL PROPERTY IREE_COMPILER_INCLUDED_PLUGIN_IDS)
-  if("${plugin_id}" IN_LIST _existing_plugin_ids)
-    message(SEND_ERROR "Plugin already registered: ${_plugin_id}")
-    return()
-  endif()
-
-  # Include it.
-  set(IREE_COMPILER_IN_ADD_PLUGIN "${plugin_id}")
-  set_property(GLOBAL APPEND PROPERTY IREE_COMPILER_INCLUDED_PLUGIN_IDS "${plugin_id}")
-  set(_binary_dir "${IREE_BINARY_DIR}/compiler/plugins/${_plugin_id}")
+  file(CONFIGURE OUTPUT "${IREE_COMPILER_PLUGINS_BINARY_DIR}/CMakeLists.txt"
```

Comment:
I'm a bit confused why we need to generate a CMakeLists.txt file that we then include rather than just including the .cmake files directly. What would be different if we just called `include("${_d}/${_f}")` in the loop above (after setting shared libs appropriately)?